File: validation/validation-annotation-processor/src/main/java/io/koraframework/validation/annotation/processor/extension/ValidKoraExtensionFactory.java
package io.koraframework.validation.annotation.processor.extension;

import io.koraframework.kora.app.annotation.processor.extension.ExtensionFactory;
import io.koraframework.kora.app.annotation.processor.extension.KoraExtension;
import io.koraframework.validation.annotation.processor.ValidMeta;

import javax.annotation.processing.ProcessingEnvironment;
import java.util.Optional;

import static io.koraframework.validation.annotation.processor.ValidTypes.VALID_TYPE;

public final class ValidKoraExtensionFactory implements ExtensionFactory {

    @Override
    public Optional<KoraExtension> create(ProcessingEnvironment processingEnvironment) {
        var element = processingEnvironment.getElementUtils().getTypeElement(VALID_TYPE.canonicalName());
        return (element == null)
            ? Optional.empty()
            : Optional.of(new ValidKoraExtension(processingEnvironment));
    }
}

File: validation/validation-symbol-processor/src/main/kotlin/io/koraframework/validation/symbol/processor/extension/ValidKoraExtensionFactory.kt
package io.koraframework.validation.symbol.processor.extension

import com.google.devtools.ksp.getClassDeclarationByName
import com.google.devtools.ksp.processing.CodeGenerator
import com.google.devtools.ksp.processing.KSPLogger
import com.google.devtools.ksp.processing.Resolver
import io.koraframework.kora.app.ksp.extension.ExtensionFactory
import io.koraframework.kora.app.ksp.extension.KoraExtension
import io.koraframework.validation.symbol.processor.ValidTypes.VALID_TYPE

class ValidKoraExtensionFactory : ExtensionFactory {
    override fun create(resolver: Resolver, kspLogger: KSPLogger, codeGenerator: CodeGenerator): KoraExtension? {
        val json = resolver.getClassDeclarationByName(VALID_TYPE.canonicalName)
        return if (json == null) {
            null
        } else {
            ValidKoraExtension()
        }
    }
}
